### PR TITLE
Fix prettier & linting; improve testing

### PR DESCRIPTION
**Pull request by lorensr**
Done:

* Add CI
* Enable tests to run without building
* Run prettier
* Fix some of the linting errors

Any tips on fixing the rest of the lint errors pictured below?

- Do we want to silence the missing function return type?
- For the `workflow.test.js`, add to `activites-examples/tsconfig` (can add to `include`, but then have to change `rootDir`) or create a new `test/tsconfig`?
- What to do for all the JS files?

(Feel free to add commits to this PR if you'd like)

image

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -103,6 +103,7 @@ dist
 # TernJS port file
 .tern-port
 
-package-lock.json
+*/package-lock.json
+*/**/package-lock.json
```
Comment (bergundy) [inline .gitignore:107]: I'm pretty sure that `package-lock.json` is enough here.

File: .prettierignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+.prettierignore
```
Comment (bergundy) [inline .prettierignore:1]: You'll probably want to ignore generated JS files as well, see the list in the sdk-node repo: https://github.com/temporalio/sdk-node/blob/main/.prettierignore

File: activities-dependency-injection/README.md [hunk 3]
```diff
@@ -7,7 +7,7 @@ This sample shows how to share dependencies between activities: for example, whe
 1. Make sure the Temporal Server is running locally. Follow the [Quick install guide](https://docs.temporal.io/docs/server/quick-install) to do that.
 2. Run `npm install` to install dependencies.
 3. Run `npm run build` to compile the project.
-3. Run `npm start.watch` to start the worker and reload on changes. Leave the worker process running.
+4. Run `npm start.watch` to start the worker and reload on changes. Leave the worker process running.
```
Comment (bergundy) [inline activities-dependency-injection/README.md:10]: This would have been solved by just using `1.` for all the list items.

File: activities-dependency-injection/src/activities.ts [hunk 4]
```diff
@@ -1,11 +1,11 @@
 // @@@SNIPSTART typescript-activity-with-deps
-export const createActivities = (dbConnection: Function) => ({
+export const createActivities = (dbConnection: () => string) => ({
```
Comment (bergundy) [inline activities-dependency-injection/src/activities.ts:2]: I fixed this in #26

**Review by bergundy — COMMENTED** [review]
Overall looks like a good start, we should test the actual samples too somehow.

File: activities-dependency-injection/README.md (hunk 3 above)
Comment (lorensr) [reply activities-dependency-injection/README.md:10]: Think this was autofixed by prettier

File: hello-world/package.json [hunk 5]
```diff
@@ -19,10 +20,13 @@
     ]
   },
   "dependencies": {
-    "temporalio": "0.5.0"
+    "temporalio": "0.4.2"
```
Comment (vkarpov15) [inline hello-world/package.json:23]: Why is this change necessary? Might conflict with #29

File: package.json [hunk 6]
```diff
@@ -2,6 +2,8 @@
   "name": "samples-node",
   "private": true,
   "scripts": {
+    "pretest": "cd activities-examples && npm install",
+    "test": "cd activities-examples && npm test",
```
Comment (vkarpov15) [inline package.json:6]: Maybe use [pushd and popd](https://fishshell.com/docs/current/cmds/pushd.html) instead so this doesn't change the user's cwd? Also, would be great if we didn't have to explicitly list all examples here. Maybe we write a test script instead?

File: package.json [hunk 7]
```diff
@@ -15,6 +17,7 @@
     "eslint-config-prettier": "^8.3.0",
     "husky": "^7.0.2",
     "lint-staged": "^11.2.3",
+    "next": "^11.1.2",
```
Comment (vkarpov15) [inline package.json:20]: Is this necessary? I don't think we use next outside of the nextjs examples.

File: package.json (hunk 6 above)
Comment (lorensr) [reply package.json:6]: Will try switching to lerna

File: hello-world/package.json (hunk 5 above)
Comment (lorensr) [reply hello-world/package.json:23]: I don't remember. It did, and it's now 0.11.x 😄

File: .shared/.eslintignore [hunk 8]
```diff
@@ -0,0 +1,3 @@
+node_modules
+lib
+.eslintrc.js
```
Comment (bergundy) [inline .shared/.eslintignore:3]: We can just lint this once and not ignore it

File: .shared/tsconfig.json [hunk 9]
```diff
@@ -2,8 +2,6 @@
   "extends": "@tsconfig/node16/tsconfig.json",
   "version": "4.4.2",
   "compilerOptions": {
-    "emitDecoratorMetadata": false,
-    "experimentalDecorators": false,
```
Comment (bergundy) [inline .shared/tsconfig.json:6]: There's no way to comment on this but we don't need `composite`.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+on:
+  pull_request:
+    branches: [$default-branch]
+  push:
+    branches: [$default-branch, next]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
```
Comment (bergundy) [inline .github/workflows/ci.yml:11]: when we do actual testing let's add mac and later windows (missing SDK support)

File: .eslintrc.json [hunk 11]
```diff
@@ -10,5 +10,8 @@
     "plugin:@typescript-eslint/eslint-recommended",
     "plugin:@typescript-eslint/recommended",
     "prettier"
-  ]
+  ],
+  "rules": {
```
Comment (bergundy) [inline .eslintrc.json:14]: Why do we need global `eslint` and `eslintignore`?

File: activities-dependency-injection/package.json [hunk 12]
```diff
@@ -7,7 +7,8 @@
     "build.watch": "tsc --build --watch",
     "start": "ts-node src/worker.ts",
     "start.watch": "nodemon src/worker.ts",
-    "workflow": "ts-node src/execute-workflow.ts"
+    "workflow": "ts-node src/execute-workflow.ts",
+    "lint": "eslint ."
```
Comment (bergundy) [inline activities-dependency-injection/package.json:11]: I wish we had a template for this too

File: activities-dependency-injection/src/worker.ts [hunk 13]
```diff
@@ -6,6 +6,7 @@ import { createActivities } from './activities';
 async function run() {
   // Mock DB connection initialization in Worker
   const db = {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (bergundy) [inline activities-dependency-injection/src/worker.ts:9]: Instead please add the following to the shared eslintrc:
```
      "@typescript-eslint/no-unused-vars": [
        1,                                                                                                                                                                                                          {
          "argsIgnorePattern": "^_"                                                                                                                                                                                 }
      ]
```

File: activities-examples/src/activities/fake-progress.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
+// @@@SNIPSTART nodejs-activity-fake-progress
```
Comment (bergundy) [inline activities-examples/src/activities/fake-progress.ts:1]: I deleted this file, should not be here

File: activities-examples/src/test/workflow.test.ts [hunk 15]
```diff
@@ -67,7 +71,11 @@ describe('example workflow', function () {
 
     await assert.rejects(
       () => workflow.execute(),
-      (err: any) => err.name === 'WorkflowExecutionFailedError' && err.cause.cause.message === 'example error'
+      (err: unknown) =>
+        err instanceof WorkflowExecutionFailedError &&
+        err.cause &&
+        'cause' in err.cause &&
+        err.cause['cause']['message'] === 'example error'
```
Comment (bergundy) [inline activities-examples/src/test/workflow.test.ts:78]: If we want to be exact use something like this:
`err.cause instanceof ActivityFailure && err.cause.cause && intstanceof ApplicationFailure && err.cause.cause.message === 'example error'`

File: ecommerce-one-click/nextjs/temporal/src/workflows/index.ts [hunk 16]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
```
Comment (bergundy) [inline ecommerce-one-click/nextjs/temporal/src/workflows/index.ts:1]: Let's not ignore the return type here, only in `createActivites`

Comment (bergundy) [reply ecommerce-one-click/nextjs/temporal/src/workflows/index.ts:1]: OIC, this is outdated

File: ecommerce-one-click/nextjs/tsconfig.json [hunk 17]
```diff
@@ -18,12 +14,6 @@
...
   },
-  "include": [
-    "next-env.d.ts",
-    "**/*.ts",
-    "**/*.tsx"
-  ],
-  "exclude": [
-    "node_modules"
-  ]
+  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
```
Comment (bergundy) [inline ecommerce-one-click/nextjs/tsconfig.json:17]: This could be a composite project with references to the worker so it's all built together.
Not for this PR though.

File: .shared/.eslintrc.js [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+    project: './tsconfig.json',
+    tsconfigRootDir: __dirname,
+  },
+  plugins: ['@typescript-eslint'],
+  extends: [
+    'eslint:recommended',
+    'plugin:@typescript-eslint/eslint-recommended',
+    'plugin:@typescript-eslint/recommended',
+    'prettier',
+  ],
```
Comment (bergundy) [inline .shared/.eslintrc.js:14]: Please add these too:
```
      "object-shorthand": ["error", "always"],
      "deprecation/deprecation": "warn",
      "@typescript-eslint/no-explicit-any": 0,
```

File: .eslintrc.json (hunk 11 above)
Comment (bergundy) [reply .eslintrc.json:14]: In `sdk-node` we have it because we don't have the per package config.
Here, I think it'd make sense to actually run eslint per-package, otherwise what's the point of having those configs?

File: .husky/.gitignore [hunk 19]
```diff
@@ -0,0 +1 @@
+_
```
[the remark this answers is not in the document]
Comment (lorensr) [reply .husky/.gitignore:1]: Think this is included by default by husky? At least I didn't add it

image

File: .husky/pre-commit [hunk 20]
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (lorensr) [reply .husky/pre-commit:1]: Looks like I can make it a node script instead

File: activities-dependency-injection/package.json (hunk 12 above)
Comment (lorensr) [reply activities-dependency-injection/package.json:11]: For package.json? I think too many differences to be worthwhile

File: .eslintrc.json (hunk 11 above)
Comment (bergundy) [reply .eslintrc.json:14]: I know, my point is that we should lint with the per-package config so users aren't surprised if they try to lint our code and and it fails.

File: activities-dependency-injection/package.json (hunk 12 above)
Comment (bergundy) [reply activities-dependency-injection/package.json:11]: I didn't mean the whole file but it would have been nice to have some shared structure that we can sync.
Definitely not for this PR.

File: package.json (hunk 6 above)
Comment (bergundy) [reply package.json]: pushd and popd don't matter here, script runs in a different shell

File: package.json (hunk 7 above)
Comment (bergundy) [reply package.json:20]: Basically I think we should run lint per sample (concurrently with lerna)

File: .husky/.gitignore (hunk 19 above)
[the remark this answers is not in the document]
Comment (bergundy) [reply .husky/.gitignore:1]: oh, only used husky once, didn't know.

File: .shared/.eslintignore (hunk 8 above)
Comment (bergundy) [reply .shared/.eslintignore:3]: oh, okay, I see why now.

File: .husky/pre-commit [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+# . "$(dirname "$0")/_/husky.sh"
+
+npx lint-staged
+
+for file in *; do
+  if [[ -d "$file" ]] # is a directory?
+  then
+    dir=$file
+    if ! [[ " node_modules ecommerce-one-click " =~ " $dir " ]]
```
[the remark this answers is not in the document]
Comment (bergundy) [reply .husky/pre-commit:10]: I'd at least just prompt and say something like:
```
Running pre-commit hook, this will overwrite any config files created in samples, proceed? [Y/n]
```

File: .husky/pre-commit (hunk 20 above)
[the remark this answers is not in the document]
Comment (bergundy) [reply .husky/pre-commit:1]: Please

[the remark this answers is not in the document]
Comment (bergundy) [reply .husky/pre-commit:1]: want to try [zx](https://github.com/google/zx)? I haven't tried it yet.

[the remark this answers is not in the document]
Comment (lorensr) [reply .husky/pre-commit:1]: In order to change `#!/bin/sh` to `#!/usr/bin/env zx`, I think all devs would need to have zx globally installed. At least it didn't work for me with it only in `node_modules`. Dunno if that would be worth it?

[the remark this answers is not in the document]
Comment (lorensr) [reply .husky/pre-commit:1]: Turned off verbosity. Output now:

```
$ npx zx .shared/copy-shared-files.mjs 
Running pre-commit hook.
This will overwrite any changes made to config files in samples (like hello-world/tsconfig.json).
Proceed? [Y/n] y
Copying config files from .shared/ to samples... done.
```

**Comment by lorensr** [comment]
Fixed otel with `npm update temporalio` and updating `workflowModules`.

File: .shared/copy-shared-files.mjs [hunk 22]
```diff
@@ -0,0 +1,65 @@
+const NON_SAMPLES = ['node_modules'];
```
Comment (bergundy) [inline .shared/copy-shared-files.mjs:1]: nit: Should this be in `.shared`? Maybe create a scripts directory?
